### ISSUE TITLE
Restart simulations don't match continuous simulations in raw_high_lat

### ISSUE THREAD

**Issue by joeschoonover** [SWPC-IPE/WAM-IPE#117]
Similar to Issue #33, I've found that a 1hr+1hr and 2hr simulation don't produce identical results on the raw_high_lat branch, even after carrying over changes from the development branch. I've verified that development achieves bit-for-bit agreement in 1hr+1hr and 2hr simulations and that data is indeed being read in from the restart files properly and passed to IPE through the mediator.

Currently, it is suspected that IPE's internal time management with the new sub-stepping required for the high latitude transport features is causing this issue. This needs to be investigated further.

**Comment by joeschoonover** [comment]
![ipe_raw_high_lat_timestepping](https://user-images.githubusercontent.com/31802478/35758524-b474874a-0831-11e8-8a0d-64809191c298.png)

This is a diagram showing what I think is supposed to happen within IPE. There is one "macro"-time-step  (dt=180 seconds), and the plasma routine is called in "micro"-time steps to satisfy interpolation constraints in the Lagrangian perpendicular advection scheme. Currently, with all the conditionals around eldyn, neutral, and plasma calls it's unclear what is being called when. Further, the sub-steps inside the stepback scheme is currently using the macro-time step (`time_step`) and not the micro time-step (`ip_freq_plasma`). 

Currently, the internal time loop is managed by a namelist parameter `internaltimeloopmax`. Instead the number of internal time steps should be `time_step` divided by `ip_freq_plasma`

**Comment by joeschoonover** [comment]
I have verified that neutral fields seen by IPE on the raw_high_lat branch are identical between restart and continuous runs. This suggests this issue is not related to #33

**Comment by joeschoonover** [comment]
I'm currently working on cleaning up the IPELIB and IPE source code in order to get a better handle on the time management within IPE. IPE standalone runs are being checked, and so far seem to be working as expected.

**Comment by joeschoonover** [comment]
ip_freq_paraTrans namelist parameter has been removed from IPE. The time stepping is now handled using `time_step`, `solar_forcing_time_step`, and `perp_transport_time_step`. 

## New time stepping scheme and parameters overview
* `time_step` is set to the same time step as WAM in coupled mode ( `DELTIM` environment variable ).

* `solar_forcing_time_step` is the time step size associated with updating forcing parameters (kp, ap, f107). Within a single forward step of IPE, the `eldyn` and `plasma` routines are called `time_step`/`solar_forcing_time_step` times; `plasma` uses a time step of `solar_forcing_time_step`. This time step size must be less that or equal to the `time_step`.

* `perp_transport_time_step` is the time step size associated with the sub-stepping needed for advancing the perpendicular transport routine. This time step must be less than or equal to the `solar_forcing_time_step`. The routine `perpendicular_transport` is called `solar_forcing_time_step`/`perp_transport_time_step` times within `module_sub_plasma` before calling the flux tube solver.

**Comment by IonospherePlasmasphereElectrodynamics** [comment]
the location of the max difference in nmf2 was identified, mp=69 lp=12

**Comment by IonospherePlasmasphereElectrodynamics** [comment]
lp=12 corresponds to magnetic latitude=77.49 [deg]

**Comment by joeschoonover** [comment]
Does this give us an idea of what the problem is ?

**Comment by IonospherePlasmasphereElectrodynamics** [comment]
it turns out that the solar wind parameters are different, which results in difference in the electric potential and electric fields. Below is the comparison of the values between the Joe's two runs at utime=3600[s]: 

(1)1hr_1hr
sw(fort.1001)
   utime    bt,      angle,  v_sw,    tilt
   3600    4.98   80.88  449.47 -0.2416E-02

poten:(fort.1002)
   utime,  MAX(pot), MIN(pot), CPCP
    3600   14.61        -28.74     43.35
(2)2hr
sw:
    3600    4.45  182.34  442.36 -0.2416E-02
poten:
    3600   26.12  -40.67   66.78

**Comment by joeschoonover** [comment]
Excellent. Which routine is responsible for reading in the solar wind parameters ? Does this occur within the call to eldyn ?

**Comment by IonospherePlasmasphereElectrodynamics** [comment]
it is suspected that the values of LPI (the index to identify the solar wind parameters at a given time, bt=swbt(LPI), for example) are different.  
LPI is calculated in eldyn/module_sub_eldyn.f:
https://github.com/SWPC-IPE/WAM-IPE/blob/raw_high_lat/IPELIB/src/eldyn/module_sub_eldyn.f#L81

**Comment by joeschoonover** [comment]
It looks like swbt is allocated in module_input_parameters and has size of f107_kp_size

**Comment by joeschoonover** [comment]
Currently, `f107_kp_size=2221`. With `f107_kp_interval=60`, I think this means we have enough forcing data to do roughly 1.5 days of historical simulation with our input file

**Comment by joeschoonover** [comment]
The forcing parameters are also read in (currently) in  `module_input_parameters`

**Comment by joeschoonover** [comment]
Note that `f107_kp_interval` should be equal to the `solar_forcing_time_step`... I'm updating this in the exglobal script.

**Comment by joeschoonover** [comment]
the namelist parameter `utime0LPI` should correspond to the time associated with the start of the `wam_input_f107_kp.txt` file. For now, I'm going to default the start time to 0UT, and attempt the 1hr+1hr and 2hr runs again.

**Comment by joeschoonover** [comment]
Setting utime0LPI=0 now causes both neutral and plasma files to differ between 1hr+1hr and 2hr runs

**Comment by joeschoonover** [comment]
Regarding my last message, the neutral/plasma differences were cause by a fix required to get standaloneGSM running. After correcting a minor mistake in exglobal_fcst_nems.sh ( commit f6873f1957c14510d27c8bd264503b0229898f44 ) we are back to only the plasma files differing. We did however, successfully reproduce the solar forcings used in eldyn and neutral between 1hr +1hr and 2hr .

This would suggest that the difference between 1hr+1hr and 2hr is not solely due to the solar wind parameters handling. 

Currently verifying this is accurate

**Comment by joeschoonover** [comment]
Verified that plasma output differs between 1hr+1hr and 2hr simulations, but the Ap, Kp, and F10.7 values in eldyn and module_neutral are identical between 1hr+1hr and 2hr simulations. Additionally, the electric potential (fort.1002 file) is identical between the 1hr+1hr and 2hr simulations.

This confirms that the treatment of forcing parameters (in these two locations) is not responsible for plasma differences.

**Comment by joeschoonover** [comment]
When comparing min/max of plasma output between the 1hr+1hr, there is a striking difference at the restart time : 
![screenshot from 2018-02-08 16-02-34](https://user-images.githubusercontent.com/31802478/36003045-9e1cc162-0ce9-11e8-88bd-9871dc5373f9.png)

**Comment by joeschoonover** [comment]
It's peculiar that some of the plasma values are negative, as indicated by a minimum that is less than zero.

**Comment by IonospherePlasmasphereElectrodynamics** [comment]
have you ever noticed the negative MIN values in the previous runs?

**Comment by joeschoonover** [comment]
We've never looked at it directly like this. Since the development branch achieves bit-for-bit reproduction between 1hr+1hr and 2hr simulations and this is the only thing that stands out on this branch as odd, I suspect these two observations are related. I'm not sure what would be gained by measuring this on development though.

**Comment by joeschoonover** [comment]
I'll be looking through all of the modules that modify the plasma_3d array this morning to see if I can figure out why we get negative values at all, and why the minimum hits zero on the start of the restart run

**Comment by IonospherePlasmasphereElectrodynamics** [comment]
plasma_3d array is a decomposed array, so a little bit of caution would be required. one would need to take MINVAL() in the SMS-SERIAL part. the negative values could be coming from the irrelevant part of the array.

**Comment by joeschoonover** [comment]
the minimum calculation is done within SMS SERIAL region. Before calling the min/max routines, appropriate exchanges are done to build a global array.

**Comment by joeschoonover** [comment]
Notice that in later times, that the maximum also differs between 1hr+1hr and 2hr simulations

**Comment by IonospherePlasmasphereElectrodynamics** [comment]
Joe, would you mind sharing your latest 2 runs(1hr+1hr vs. 2hr)? I would like to look into the results and compare in details. Thanks.

**Comment by joeschoonover** [comment]
Any work currently being done on raw_high_lat should migrate to raw_high_lat_v2.

Locally, if you are working on raw_high_lat, temporarily save any work in progress
```
git stash
```
Then, check out raw_high_lat_v2
```
git fetch
git checkout origin/raw_high_lat_v2 -b raw_high_lat_v2
```
Then, reapply any of your "work in progress"
```
git stash apply
```

Note that the git stash commands only need to be used if you have uncommitted work in progress.

**Comment by joeschoonover** [comment]
Need to determine which plasma parameter differs between the two simulations.

**Comment by akubaryk** [comment]
We talked about this briefly during the developer's meeting today. As a brief followup, wanted to confirm that it wasn't a pre-processing issue. From `scripts/interpolate_input_parameters`:

```
$ ./interpolate_input_parameters.py -d 38 -s `ndate -36 2013031600` -p /scratch3/NCEPDEV/swpc/noscrub/Adam.Kubaryk/WAM-IPE_INPUT_PARAMETERS
$ mv wam_input_f107_kp.txt 2h.txt
$ ./interpolate_input_parameters.py -d 37 -s `ndate -36 2013031600` -p /scratch3/NCEPDEV/swpc/noscrub/Adam.Kubaryk/WAM-IPE_INPUT_PARAMETERS
$ mv wam_input_f107_kp.txt 1h.txt
$ ./interpolate_input_parameters.py -d 37 -s `ndate -36 2013031601` -p /scratch3/NCEPDEV/swpc/noscrub/Adam.Kubaryk/WAM-IPE_INPUT_PARAMETERS
$ mv wam_input_f107_kp.txt +1h.txt
```

So `2h.txt` would contain the normal 2 hour run (36 hours of past data + 2 hours), `1h.txt` should contain the 2 hour run except for the last hour of the file (2013031601-2013031602), and `+1h.txt` should contain the 2 hour run except for the first hour of the file (2013031412-2013031413). And, indeed, diffing the files against one another shows exactly that.

`F107_KP_SIZE` and (thanks to @ZhuxiaoLi66) `F107_KP_DATA_SIZE` are defined by the length of `wam_input_f107_kp.txt`, so those should be fine. `F107_KP_SKIP_SIZE` is set so that IPE starts from 2161 (36 hours*60) for the first timestep and MSIS starts from 1 -- the 24h Kp average at 36h ago. `F107_KP_INTERVAL` is now hard-coded for the minute-cadence. These are all of the relevant namelist options, leaving model code as the other option.

TL;DR: If the issue is with the solar/geomagnetic input parameters, it must be from array indexing in IPE. Hopefully using fixed values with the new pull request can help us track down the cause.
